apps/onnx/model.cpp

```diff
@@ -0,0 +1,549 @@
+        break;
+    }
+    case onnx::TensorProto::UINT32: {
+        prepare_input<uint32_t>(pipeline, input_name, input_shape);
+        break;
+    }
+    case onnx::TensorProto::UINT64: {
+        prepare_input<uint64_t>(pipeline, input_name, input_shape);
+        break;
+    }
+    case onnx::TensorProto::FLOAT: {
+        prepare_input<float>(pipeline, input_name, input_shape);
+        break;
+    }
+    case onnx::TensorProto::DOUBLE: {
+        prepare_input<double>(pipeline, input_name, input_shape);
+        break;
+    }
+    default: {
+        throw std::invalid_argument(
+            std::string("Unsupported type for input ") + input_name);
+    }
+    }
+}
+
+template<typename T>
+py::array_t<T> export_output(
+    const Halide::Buffer<T> &output_values,
+    const HalideModel &pipeline,
+    int output_id) {
+    Tensor node = pipeline.model->outputs.at(pipeline.output_names.at(output_id));
+    std::vector<int> output_shape;
+    int stride = 1;
+    const int rank = node.shape.size();
+    std::vector<int> np_strides;
+    for (int i = 0; i < rank; ++i) {
+        const int64_t *dim = Halide::Internal::as_const_int(node.shape[i]);
+        if (!dim) {
+            throw std::invalid_argument(
+                "Couldn't statically infer dim " + std::to_string(i) + " of output " +
+                std::to_string(output_id));
+        }
+        output_shape.push_back(*dim);
+        np_strides.push_back(stride);
+        dim = Halide::Internal::as_const_int(node.shape[rank - (i + 1)]);
+        if (!dim) {
+            throw std::invalid_argument(
+                "Couldn't statically infer dim " + std::to_string(rank - (i + 1)) +
+                " of output " + std::to_string(output_id));
+        }
+        stride *= *dim;
+    }
+    std::reverse(np_strides.begin(), np_strides.end());
+
+    // TODO:Better handeling of scalar outputs.
+    if (output_shape.size() > 0) {
+        py::array_t<T, py::array::c_style> result(output_shape);
+
+        T *mutable_data = result.mutable_data();
+
+        output_values.for_each_element([&](const int *halide_coords) {
+            int np_index = 0;
+            for (int i = 0; i < rank; i++) {
+                np_index += halide_coords[i] * np_strides[i];
+            }
+            mutable_data[np_index] = output_values(halide_coords);
+        });
+        return result;
+    } else {
+        py::array_t<T, py::array::c_style> result({ 1 });
+
+        T *mutable_data = result.mutable_data();
+
+        output_values.for_each_element([&](const int *halide_coords) {
+            int np_index = 0;
+            for (int i = 0; i < rank; i++) {
+                np_index += halide_coords[i] * np_strides[i];
+            }
+            mutable_data[np_index] = output_values(halide_coords);
+        });
+        return result;
+    }
+}
+
+static Halide::Type onnx_type_to_halide_type(int t) {
+    switch (t) {
+    case onnx::TensorProto::FLOAT:
```

Comment:
This isn't essential for this PR to land, but we have an awful lot of repeated code that maps type-to-type; is there any way this could be consolidated somehow?

Reply:
Agree this and get_tesnsor_type(Tensor& tensor) should be consolidated. Adding a TODO.